drivers/char/broadcom/vc_mem.c

```diff
@@ -157,6 +164,115 @@ vc_mem_compat_ioctl(struct file *file, unsigned int cmd, unsigned long arg)
 }
 #endif
 
+static loff_t
+vc_mem_llseek(struct file *file, loff_t off, int whence) {
+	switch (whence) {
+		case SEEK_CUR:
+			off += file->f_pos;
+			// fallthrough
+		case SEEK_SET:
+			file->f_pos = off;
+			break;
+		default:
+			return -EINVAL;
+	}
+
+	force_successful_syscall_return();
+	return off;
+}
+
+static ssize_t
+vc_mem_read(struct file *file, char __user *read_buf, size_t read_size, loff_t *ppos) {
+	int err = 0;
+	void *buf;
+	dma_addr_t buf_paddr;
+	dma_addr_t src_begin, src_end, src_at;
+
+	struct vc_mem *drv = file->private_data;
+	struct device *dev = drv->misc.parent;
+	struct dma_chan *dma_chan = drv->dma_chan;
+	struct dma_device *dma_dev = dma_chan->device;
+
+	unsigned long init_pos = *ppos;
+	const unsigned long alias_cached = 0x00000000;
+	const unsigned long alias_direct = 0xc0000000;
+	const unsigned long alias_mask = 0xc0000000;
+
+	// Check the alias bits are permitted
+	if ((init_pos & alias_mask) != alias_cached &&
+	    (init_pos & alias_mask) != alias_direct &&
+	    (init_pos & alias_mask) != (drv->phys_addr & alias_mask))
+		return -EINVAL;
+
+	// Discard the cache alias bits for bounds checking
+	init_pos &= ~alias_mask;
+
+	if (init_pos < drv->base || init_pos > drv->size ||
+	    init_pos + read_size > drv->size)
+		return -EINVAL;
+
+	if (!read_size)
+		return 0;
+
+	// Use the preferred alias for accesses
+	src_begin = init_pos + drv->phys_addr;
+	src_end = src_begin + read_size;
+	src_at = src_begin;
+
+	buf = dma_alloc_coherent(dev, VC_MEM_DMA_SIZE, &buf_paddr, GFP_KERNEL);
```

Comment:
I think this will be very broken if you don't have `dma-ranges`. From what I remember on my Pi 4 it used to allocate memory anywhere (in the first 4G because that's how much I had), so the DMA engine ended up overwriting the wrong memory. Funnily enough that killed some pages in the page cache that were then written back so I lost some files.

Reply:
Several of these issues stem from putting the vc_mem node under /axi, forcing one to duplicate the ranges and dma-ranges of /soc. We can replace the dev in the dma_* calls with the DMA controller, but it seems more correct to move vc_mem under /soc and locate /axi/reg explicitly.

Reply:
Good idea. Should we add a pointer from `/soc/vc_mem` to `/axi/vc_mem`? Something like `/soc/vc_mem { raspberrypi,legacy-node = <&axi_vc_mem>; };`